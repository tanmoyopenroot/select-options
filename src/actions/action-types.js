/**
 * Action types for redux actions
 * @module actions/action-types
 */

export const FETCH_DATA = 'FETCH_DATA';
export const FETCHED_USERS_DATA = 'FETCHED_USERS_DATA';
export const FETCHED_ROLES_DATA = 'FETCHED_ROLES_DATA';
export const FETCHED_PROJECT_DATA = 'FETCHED_PROJECT_DATA';

export const SELECTED_DATA = 'SELECTED_DATA';

export const SET_SELECTED_USER = 'SET_SELECTED_USER';
export const SET_SELECTED_ROLE = 'SET_SELECTED_ROLE';
export const SET_SELECTED_PROJECT = 'SET_SELECTED_PROJECT';
