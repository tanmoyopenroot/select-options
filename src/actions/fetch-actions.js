import { FETCH_DATA } from './action-types';

export const fetchData = () => ({
  type: FETCH_DATA
})
