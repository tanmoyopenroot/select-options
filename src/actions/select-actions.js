import { SELECTED_DATA } from './action-types';

export const saveSelectData = (type, set, id) => ({
  type: SELECTED_DATA,
  payload: {
    type,
    setID: set,
    id,
  }
})
