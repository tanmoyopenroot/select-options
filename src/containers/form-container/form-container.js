/**
 * FormContainer
 * @module containers/form-container
 */

import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import SetContainer from '../set-container';
import { getUsers } from '../../selectors/users-selector';
import { getRoles } from '../../selectors/roles-selector';
import { getProjects } from '../../selectors/projects-selector';
import { checkError } from '../../selectors/error-selector';
import { getSelectedData } from '../../selectors/selected-selector';
import { fetchData } from '../../actions/fetch-actions'
import styled from 'styled-components';

const SubmitButton = styled.button`
  padding: 20px 40px;
  width: 200px;
  border: none;
  border-radius: 2px;
  background-color: #388E3C;
  color: #fff;
  text-transform: uppercase;
  font-weight: bold;
  font-size: 16px;
  margin: 30px auto;
  display: block;
`;

const ErrorContainer = styled.div`
  color: #B71C1C;
  text-align: center;
  margin-top: 20px;
`;

const OutputContainer = styled.div`
  text-align: center;
  background-color: #fff;
  padding: 20px;
`;

/**
 * @class FromContainer
 * @extends {React.Component}
 */

export class FromContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      errMsg: '',
      showErr: false,
      submitOutput: '',
      showOutput: false,
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.doublePresent = this.doublePresent.bind(this);
  }

  componentDidMount() {
    const { fetchData } = this.props;
    fetchData();
  }

  processDataforOutput() {
    const { selected, users, roles, projects } = this.props;

    let data = {};
    Object.keys(selected.users).map(set => (
      data[set] = {
        user: users.usersHash[selected.users[set]],
      }
    ));

    Object.keys(selected.roles).map(set => (
      data[set] = {
        ...data[set],
        role: roles.rolesHash[selected.roles[set]],
      }
    ));

    Object.keys(selected.projects).map(set => (
      data[set] = {
        ...data[set],
        project: projects.projectsHash[selected.projects[set]],
      }
    ));

    this.setState({
      submitOutput: JSON.stringify(data),
      showOutput: true,
    })
    // console.log(data);
  }

  allSelected() {
    const { selected, users, roles, projects } = this.props;
    if (
      Object.keys(selected.users).length === (users.usersID).length &&
      Object.keys(selected.roles).length === (roles.rolesID).length &&
      Object.keys(selected.projects).length === (projects.projectsID).length
    ) {
      return true;
    }

    return false;
  }

  doublePresent(obj) {
    let flag = {};
    for (let id in  obj) {
      if (!flag[obj[id]]) {
        flag[obj[id]] = true
      } else {
        return true
      }
    }
    return false;
  }

  _showForm() {
    const { showErr, errMsg, showOutput, submitOutput } = this.state;
    const { users } = this.props;
    const { usersID } = users;

    return (
      <Fragment>
        {
          usersID.map(id => (
            <SetContainer
              key={`set-${id}`}
              setID={id}
            />
          ))
        }
        {
          showErr ?
            <ErrorContainer>
              { errMsg }
            </ErrorContainer>
          :
            null
        } 
        <SubmitButton
          onClick={this.handleSubmit}
        >
          Submit
        </SubmitButton>
        {
          showOutput && !showErr?
            <OutputContainer>
              {submitOutput}
            </OutputContainer>
          :
            null
        }
      </Fragment>
    )
  }

  handleSubmit() {
    const { selected } = this.props;
    const { users, roles, projects } = selected;

    if (!this.allSelected()) {
      this.setState({
        errMsg: 'You need to select the options before submit.',
        showErr: true,
      })
      return;
    }

    let userDoublePresent = this.doublePresent(users);
    let rolesDoublePresent = this.doublePresent(roles);
    let projectDoublePresent = this.doublePresent(projects);

    if (userDoublePresent || rolesDoublePresent || projectDoublePresent) {
      this.setState({
        errMsg: 'Each user can be assigned just one role per project.',
        showErr: true,
      })
      return;
    }

    this.setState({
      showErr: false,
    })

    this.processDataforOutput();
  }

  render () {
    const { fetchError } = this.props;
    const { error, message } = fetchError;

    return (
      <Fragment>
        {
          error ?
            <ErrorContainer>
              {message}
            </ErrorContainer>
          :
            this._showForm()
        }
      </Fragment>
    );
  }
}

FromContainer.propTypes = {
  users: PropTypes.object,
  roles: PropTypes.object,
  projects: PropTypes.object,
  fetchError: PropTypes.object,
  selected: PropTypes.object,
  fetchData: PropTypes.func,
}

const mapStateToProps = state => ({
  users: getUsers(state),
  roles: getRoles(state),
  projects: getProjects(state),
  fetchError: checkError(state),
  selected: getSelectedData(state),
})

const mapDispatchToProps = {
  fetchData,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FromContainer);
