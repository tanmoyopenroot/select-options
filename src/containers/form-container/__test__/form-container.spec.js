import React from 'react';
import { FromContainer } from '../form-container';
import { shallow } from 'enzyme';

import { fetchData } from '../../../actions/fetch-actions';

const setup = () => {
  const output = {
    set_1: {
      user: {
        id: 2,
        name: 'Alice',
      },
      role: {
        id: 3,
        name: 'Viewer',
      },
      project: {
        id: 1,
        name: 'Trip to space',
      }
    },
    set_2: {
      user: {
        id: 1,
        name: 'John Doe',
      },
      role: {
        id: 2,
        name: 'Editor',
      },
      project: {
        id: 2,
        name: 'Assembly Ikea furniture',
      }
    },
    set_3: {
      user: {
        id: 3,
        name: 'Bob',
      },
      role: {
        id: 1,
        name: 'Admin',
      },
      project: {
        id: 3,
        name: 'Datumize Zentral',
      }
    }
  }

  const selected = {
    projects: {
      set_1: 1,
      set_2: 2,
      set_3: 3,
    },
    roles: {
      set_1: 3,
      set_2: 2,
      set_3: 1,
    },
    users: {
      set_1: 2,
      set_2: 1,
      set_3: 3,
    },
  }

  const users = {
    usersID: [1, 2, 3],
    usersHash: {
      1: { id: 1, name: 'John Doe'}, 
      2: { id: 2, name: 'Alice' },
      3: { id: 3, name: 'Bob' }
    },
  }
  
  const roles = {
    rolesID: [1, 2, 3],
    rolesHash: {
      1: { id: 1, name: 'Admin'}, 
      2: { id: 2, name: 'Editor' },
      3: { id: 3, name: 'Viewer' }
    },
  }
  
  const projects = {
    projectsID: [1, 2, 3],
    projectsHash: {
      1: { id: 1, name: 'Trip to space'},
      2: { id: 2, name: 'Assembly Ikea furniture' },
      3: { id: 3, name: 'Datumize Zentral' }
    },
  }

  const fetchError = {
    error: false,
    message: '',
  }

  return {
    users,
    roles,
    projects,
    output,
    selected,
    fetchError,
  }
} 

describe('Testing <FormContainer />', () => {
  const {
    users,
    roles,
    projects,
    selected,
    output,
    fetchError,
  } = setup();
    
  const wrapper = shallow(
    <FromContainer
      users={users}
      roles={roles}
      projects={projects}
      selected={selected}
      fetchData={fetchData}
      fetchError={fetchError}
    />,
  )

  test('Testing FormContainer\'s props', () => {
    expect(wrapper.length).toEqual(1);
    expect(wrapper.children().length).toBe(4);
    expect(Object.keys(wrapper.instance().props).length).toEqual(6);
    expect(wrapper.instance().props.users).toEqual(users);
    expect(wrapper.instance().props.roles).toEqual(roles);
    expect(wrapper.instance().props.projects).toEqual(projects);
    expect(wrapper.instance().props.fetchError).toEqual(fetchError);
    expect(wrapper).toMatchSnapshot();
  })

  test('Testing FormContainer\'s functions', () => {
    wrapper.instance().handleSubmit()
    expect(wrapper.state().errMsg).toEqual('');
    expect(wrapper.state().showErr).toEqual(false);
    expect(wrapper.state().submitOutput).toEqual(JSON.stringify(output));
    expect(wrapper.state().showOutput).toEqual(true);
  });

  test('Testing FormContainer\'s FetchError', () => {
    const setPropValue = {
      fetchError: {
        error: true,
        message: 'Error while fetching users data.'
      }
    }
    wrapper.setProps(setPropValue);
    expect(wrapper.children().length).toBe(1);
  })
})
