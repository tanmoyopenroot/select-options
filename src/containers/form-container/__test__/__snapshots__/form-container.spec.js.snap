// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <FormContainer /> Testing FormContainer's FetchError 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FromContainer
    fetchData={[Function]}
    fetchError={
        Object {
            "error": false,
            "message": "",
          }
    }
    projects={
        Object {
            "projectsHash": Object {
              "1": Object {
                "id": 1,
                "name": "Trip to space",
              },
              "2": Object {
                "id": 2,
                "name": "Assembly Ikea furniture",
              },
              "3": Object {
                "id": 3,
                "name": "Datumize Zentral",
              },
            },
            "projectsID": Array [
              1,
              2,
              3,
            ],
          }
    }
    roles={
        Object {
            "rolesHash": Object {
              "1": Object {
                "id": 1,
                "name": "Admin",
              },
              "2": Object {
                "id": 2,
                "name": "Editor",
              },
              "3": Object {
                "id": 3,
                "name": "Viewer",
              },
            },
            "rolesID": Array [
              1,
              2,
              3,
            ],
          }
    }
    selected={
        Object {
            "projects": Object {
              "set_1": 1,
              "set_2": 2,
              "set_3": 3,
            },
            "roles": Object {
              "set_1": 3,
              "set_2": 2,
              "set_3": 1,
            },
            "users": Object {
              "set_1": 2,
              "set_2": 1,
              "set_3": 3,
            },
          }
    }
    users={
        Object {
            "usersHash": Object {
              "1": Object {
                "id": 1,
                "name": "John Doe",
              },
              "2": Object {
                "id": 2,
                "name": "Alice",
              },
              "3": Object {
                "id": 3,
                "name": "Bob",
              },
            },
            "usersID": Array [
              1,
              2,
              3,
            ],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Unknown>
        <Connect(SetContainer)
                setID={1}
        />
        <Connect(SetContainer)
                setID={2}
        />
        <Connect(SetContainer)
                setID={3}
        />
        <styled.button
                onClick={[Function]}
        >
                Submit
        </styled.button>
</Unknown>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          Array [
            <Connect(SetContainer)
              setID={1}
/>,
            <Connect(SetContainer)
              setID={2}
/>,
            <Connect(SetContainer)
              setID={3}
/>,
          ],
          null,
          <styled.button
            onClick={[Function]}
>
            Submit
</styled.button>,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "set-1",
          "nodeType": "class",
          "props": Object {
            "setID": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "set-2",
          "nodeType": "class",
          "props": Object {
            "setID": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "set-3",
          "nodeType": "class",
          "props": Object {
            "setID": 3,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Submit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Submit",
          "type": [Function],
        },
        null,
      ],
      "type": Symbol(react.fragment),
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Unknown>
          <Connect(SetContainer)
                    setID={1}
          />
          <Connect(SetContainer)
                    setID={2}
          />
          <Connect(SetContainer)
                    setID={3}
          />
          <styled.button
                    onClick={[Function]}
          >
                    Submit
          </styled.button>
</Unknown>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            Array [
              <Connect(SetContainer)
                setID={1}
/>,
              <Connect(SetContainer)
                setID={2}
/>,
              <Connect(SetContainer)
                setID={3}
/>,
            ],
            null,
            <styled.button
              onClick={[Function]}
>
              Submit
</styled.button>,
            null,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "set-1",
            "nodeType": "class",
            "props": Object {
              "setID": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "set-2",
            "nodeType": "class",
            "props": Object {
              "setID": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "set-3",
            "nodeType": "class",
            "props": Object {
              "setID": 3,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Submit",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Submit",
            "type": [Function],
          },
          null,
        ],
        "type": Symbol(react.fragment),
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`Testing <FormContainer /> Testing FormContainer's props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FromContainer
    fetchData={[Function]}
    fetchError={
        Object {
            "error": false,
            "message": "",
          }
    }
    projects={
        Object {
            "projectsHash": Object {
              "1": Object {
                "id": 1,
                "name": "Trip to space",
              },
              "2": Object {
                "id": 2,
                "name": "Assembly Ikea furniture",
              },
              "3": Object {
                "id": 3,
                "name": "Datumize Zentral",
              },
            },
            "projectsID": Array [
              1,
              2,
              3,
            ],
          }
    }
    roles={
        Object {
            "rolesHash": Object {
              "1": Object {
                "id": 1,
                "name": "Admin",
              },
              "2": Object {
                "id": 2,
                "name": "Editor",
              },
              "3": Object {
                "id": 3,
                "name": "Viewer",
              },
            },
            "rolesID": Array [
              1,
              2,
              3,
            ],
          }
    }
    selected={
        Object {
            "projects": Object {
              "set_1": 1,
              "set_2": 2,
              "set_3": 3,
            },
            "roles": Object {
              "set_1": 3,
              "set_2": 2,
              "set_3": 1,
            },
            "users": Object {
              "set_1": 2,
              "set_2": 1,
              "set_3": 3,
            },
          }
    }
    users={
        Object {
            "usersHash": Object {
              "1": Object {
                "id": 1,
                "name": "John Doe",
              },
              "2": Object {
                "id": 2,
                "name": "Alice",
              },
              "3": Object {
                "id": 3,
                "name": "Bob",
              },
            },
            "usersID": Array [
              1,
              2,
              3,
            ],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Unknown>
        <Connect(SetContainer)
                setID={1}
        />
        <Connect(SetContainer)
                setID={2}
        />
        <Connect(SetContainer)
                setID={3}
        />
        <styled.button
                onClick={[Function]}
        >
                Submit
        </styled.button>
</Unknown>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          Array [
            <Connect(SetContainer)
              setID={1}
/>,
            <Connect(SetContainer)
              setID={2}
/>,
            <Connect(SetContainer)
              setID={3}
/>,
          ],
          null,
          <styled.button
            onClick={[Function]}
>
            Submit
</styled.button>,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "set-1",
          "nodeType": "class",
          "props": Object {
            "setID": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "set-2",
          "nodeType": "class",
          "props": Object {
            "setID": 2,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "set-3",
          "nodeType": "class",
          "props": Object {
            "setID": 3,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Submit",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Submit",
          "type": [Function],
        },
        null,
      ],
      "type": Symbol(react.fragment),
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Unknown>
          <Connect(SetContainer)
                    setID={1}
          />
          <Connect(SetContainer)
                    setID={2}
          />
          <Connect(SetContainer)
                    setID={3}
          />
          <styled.button
                    onClick={[Function]}
          >
                    Submit
          </styled.button>
</Unknown>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            Array [
              <Connect(SetContainer)
                setID={1}
/>,
              <Connect(SetContainer)
                setID={2}
/>,
              <Connect(SetContainer)
                setID={3}
/>,
            ],
            null,
            <styled.button
              onClick={[Function]}
>
              Submit
</styled.button>,
            null,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "set-1",
            "nodeType": "class",
            "props": Object {
              "setID": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "set-2",
            "nodeType": "class",
            "props": Object {
              "setID": 2,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "set-3",
            "nodeType": "class",
            "props": Object {
              "setID": 3,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          null,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Submit",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Submit",
            "type": [Function],
          },
          null,
        ],
        "type": Symbol(react.fragment),
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
